---
title: "Overview of the Data-set"
author: "Raphaelle Luisier, Idiap Research Institute"
date: "February 2023"
output:
  output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache =TRUE)
knitr::opts_chunk$set(fig.width=16, fig.height=10, figure.align = "center") 
library(ggplot2) # devtools::install_github('hadley/ggplot2')
library(RColorBrewer)
library(wesanderson)
library(knitr)
library(mclust)
#library(limma)
require(gplots)
#library(lme4)
#library(nlme)
#source("http://bioconductor.org/biocLite.R")
#require("GO.db")
#require("limma")
#require("topGO")
#require("biomaRt")
#require("org.Rn.eg.db")
#library(grDevices)
library(Rsamtools)
library(IRanges)
library(GenomicRanges)
library(rtracklayer)
library(geneplotter)
library("ape")
require("multtest")
require("mclust")

rm(list = ls())
```

### Load the data
3' UTR isoform quantification from 8 samples originating from rat neuronal sympathetic cell cultures where the cell body is treated with low does of NGF and distal axons are either exposed to NGF or NT3. There are a total of 8 samples: cell body versus distal axons in NGF versus NT3 culture condition across 2 technical replicates.

```{r load_data_load_scripts,echo=FALSE}
load("./data/myfiles.RData")#"anno_tot","myUTR"
#myUTR        <- import.gff("./data/myUTR.sorted.gtf",format="gtf")
#anno_tot     <- read.delim("./data/anno_tot.tab",header=TRUE,sep="\t")
names(myUTR)  <- as.character(myUTR$ID)
source("./scripts/nested_fun_differential_PAS.R")
```



### Analysis of the distribution of the raw count
The distributions of the expression of 12'771 3' UTR isoforms detected in cell body and/or distal axons of sympathetic neurons. As shown below, the detected expression levels in axonal compartment is systematically lower than in cell body compartment. As discussed in materials and method section, the amount of mRNA material obtained from axonal compartment is much lower. Furthermore the large amount of mithondria located in axonal compartments arise in a library maily composed of mithocondrial mRNAs. This leads to technical challenges in library preparation which were overcomed by performing two rounds of mRNA linear amplification. We alos find that one of the axonal samples exposed to NT3 was of lower quality (exhibiting systematically lower expression level), which we take into account in our analysis.


```{r analysis_distributions,echo=FALSE,fig.width=10, fig.height=5,warning=FALSE}
data     <- anno_tot[,grep(".raw.corrected",colnames(anno_tot))]#Raw count corrected for reads arising from anti-sense transcripts which have leaked 
row_info <- anno_tot

#Here focus on the isoforms annotated in Rn5
data.sub              <- data
data.sub              <- data[row_info$is.conservative==TRUE,]
data.log2             <- log2(data.sub)
data.log2[data.sub<=1]<-0
colnames(data.log2)   <- gsub(colnames(data.log2),pattern="\\.raw\\.corrected",repl="")
temp                  <- lapply(colnames(data.log2),function(Z)return(strsplit(Z,split="\\.")[[1]]))         
col_info              <- data.frame(do.call(args=temp,what=rbind))
colnames(col_info)    <- c("treatment","compartment","replicate")
col_info$group        <- factor(paste(col_info$treatment,col_info$compartment,sep="_"),levels=c("NGF_cb","NGF_axon","NT3_cb","NT3_axon"))
mycols_treatment      <- c("#81A4D6","#B072B0")
names(mycols_treatment)<- c("NGF","NT3")



#Look at the distribution of the data
par(mfrow=c(1,2))
boxplot(data.log2,outline=FALSE,las=2)
multidensity(data.log2)

```