---
title: "Analysis of axonal remodelling in different neurotrophic conditions"
author: "Raphaelle Luisier, Idiap Research Institute"
date: "February 2023"
output:
  html_document:
    theme: paper
#paper readable
    highlight: monochrome
    code_folding: hide
    toc: true
    toc_depth: 2
#    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache =TRUE)
knitr::opts_chunk$set(fig.width=16, fig.height=10, figure.align = "center") 
require("edgeR")
require("GO.db")
require("limma")
require("topGO")
require("biomaRt")
require("org.Rn.eg.db")
library(grDevices)
library(Rsamtools)
library(IRanges)
library(GenomicRanges)
library(rtracklayer)
library(geneplotter)
require("multtest")
require("mclust")

rm(list = ls())
```

# Load the and filter data {.tabset}
3' UTR isoform quantification from 8 samples originating from rat neuronal sympathetic cell cultures where the cell body is treated with low does of NGF and distal axons are either exposed to NGF or NT3. There are a total of 8 samples: cell body versus distal axons in NGF versus NT3 culture condition across 2 technical replicates.

```{r load_data_load_scripts}
#load("./data/myfiles.RData")#"anno_tot","myUTR"
#myUTR        <- import.gff("./data/myUTR.sorted.gtf",format="gtf")
#anno_tot     <- read.delim("./data/anno_tot.tab",header=TRUE,sep="\t")
#names(myUTR)  <- as.character(myUTR$ID)
#source("./scripts/nested_fun_differential_PAS.R")

load("~/Desktop/DataAnalsyisRiccio/NGFvsNT3/files_analysis.RData")
coloi                <- c("NGF.axon.1.raw.corrected","NGF.axon.2.raw.corrected","NT3.axon.1.raw.corrected","NT3.axon.2.raw.corrected")
myUTR        <- import.gff("~/Desktop/DataAnalsyisRiccio/Dec2016/utrid/APA_stringent/Ltot_sub_2017.gtf",format="gtf")
names(myUTR) <- as.character(myUTR$ID)
```

In order to perform 3' UTR remodelling analysis, the data-set needs to be filtered as per to remove transcripts which have only one detected isoform for example.

```{r filter_data}
# A. Remove those transcript ID which have only one isoforms
anno_tot     <- anno_tot[anno_tot$NGF.axon.is.expressed.iso|anno_tot$NT3.axon.is.expressed.iso,]
temp         <- as.data.frame(table(as.character(anno_tot$txID)))
no.iso       <- temp[match(anno_tot$txID,as.character(temp[,1])),2]
anno_tot     <- anno_tot[anno_tot$no.iso>1,]#9'686 rows   
names(myUTR) <- myUTR$ID
myUTR        <- myUTR[match(anno_tot$uniqueID,names(myUTR)),]




# B. Check that the closest neighbour is not within 300 nt
tempL                     <- anno_tot$newL
names(tempL)              <- anno_tot$uniqueID
tempG                     <- as.factor(as.character(anno_tot$txID))
myord                     <- tapply(tempL,INDEX=tempG,function(x)return(cbind(names(x)[sort(as.numeric(as.character(x)),index.return=T,decreasing=F)$ix],c(1:length(x)))))
test                      <- do.call(what=rbind,args=myord)
iso_ordered               <- as.numeric(test[match(anno_tot$uniqueID,test[,1]),2])
#nextIsoform
tempid                    <- paste(anno_tot$txID,iso_ordered,sep=".")
next.iso                  <- paste(anno_tot$txID,(as.numeric(as.character(iso_ordered))+1),sep=".")
nextID                    <- anno_tot$uniqueID[match(next.iso,tempid)]
#distTonext
distonext                 <- rep(NA,nrow(anno_tot))
sel                       <- !is.na(nextID)
d1                        <- as.numeric(as.character(anno_tot$newL))[sel]
d2                        <- as.numeric(as.character(anno_tot$newL))[match(nextID[sel],anno_tot$uniqueID)]
distonext[sel]            <- d2-d1
#Fproxi
Fproxi                    <- which(distonext<300)#none; ok what expected
Fproxi


# C. Remake ordering and imID
tempL                     <- anno_tot$newL
names(tempL)              <- anno_tot$uniqueID
tempG                     <- as.factor(as.character(anno_tot$txID))
myord                     <- tapply(tempL,INDEX=tempG,function(x)return(cbind(names(x)[sort(as.numeric(as.character(x)),index.return=T,decreasing=F)$ix],c(1:length(x)))))
test                      <- do.call(what=rbind,args=myord)
test                      <- test[match(anno_tot$uniqueID,test[,1]),]
temp                      <- as.data.frame(table(as.character(anno_tot$txID)))
no.iso                    <- temp[match(anno_tot$txID,as.character(temp[,1])),2]
anno_tot$no.iso           <- no.iso
anno_tot$iso_ordered      <- test[,2]
tempL                     <- as.character(anno_tot$iso_ordered)
names(tempL)              <- as.character(anno_tot$uniqueID)
imID1                     <- tapply(tempL,INDEX=as.factor(as.character(anno_tot$txID)),function(x)return(names(x)[x=="1"]))
imIDp                     <- as.character(imID1[match(anno_tot$txID,names(imID1))])
anno_tot$imIDp            <- imIDp

# D. Get the selection on which to focus for the analysis
# At least one of the pair (distal or proximal) must be detected in axons of NGF
sela11                 <- (anno_tot$NGF.axon.is.expressed.iso|anno_tot$NGF.axon.is.expressed.iso[match(anno_tot$imIDp,anno_tot$uniqueID)])
# At least one of the  pair (distal or proximal) must be detected in axons of NT3
sela21                 <- (anno_tot$NT3.axon.is.expressed.iso|anno_tot$NT3.axon.is.expressed.iso[match(anno_tot$imIDp,anno_tot$uniqueID)])
# Remove all the isoform I0
sela3                  <- anno_tot$uniqueID!=anno_tot$imIDp
#Proximal is expressed in at least one of the 2 axons samples 
sela41                 <- (anno_tot$NGF.axon.is.expressed.iso|anno_tot$NT3.axon.is.expressed.iso)[match(anno_tot$imIDp,anno_tot$uniqueID)]
#Distal is expressed in at least one of the 2 axons samples
sela51                 <- (anno_tot$NGF.axon.is.expressed.iso|anno_tot$NT3.axon.is.expressed.iso)
sela1                  <- sela11&sela21&sela3&sela41&sela51
subanno1               <- anno_tot[sela1,]#2'881 pairs
```


# Analysis of changes in 3' UTR usage between cell body and axons {.tabset}
We perform this analysis in each neurotrophin condition.

```{r apa_analysis}
# A.    Proximal to distal site usage AND fisher test
coloi                <- c("NGF.axon.1.raw.corrected","NGF.axon.2.raw.corrected","NT3.axon.1.raw.corrected","NT3.axon.2.raw.corrected")
ix1                  <- c(1:nrow(anno_tot))
ix2                  <- match(as.character(anno_tot$imIDp),as.character(anno_tot$uniqueID))

#A.1 RUD
mydat                <- anno_tot[,match(coloi,colnames(anno_tot))]
for(i in c(1:ncol(mydat))){mydat[,i]<-log2(1+mydat[,i])}
rownames(mydat)      <- anno_tot$uniqueID
myProximal           <- apply(mydat,2,function(x)return(x[ix2]))
myDistal             <- apply(mydat,2,function(x)return(x[ix1]))
rownames(myProximal) <-rownames(myDistal)<-anno_tot$uniqueID[ix1]
RUD                  <- do.call(lapply(c(1:4),function(x)return(myProximal[,x]-myDistal[,x])),what=cbind)
rownames(RUD)        <- anno_tot$uniqueID[ix1]
colnames(RUD)        <- unlist(lapply(colnames(mydat),function(Z)return(gsub(Z,pattern="\\.raw",repl=""))))
mRUD                  <- t(apply(RUD,1,function(x)return(tapply(x,INDEX=factor(c("NGF.axon","NGF.axon","NT3.axon","NT3.axon")),FUN=mean))))
dRUD                  <- mRUD[,1]-mRUD[,2]#NGF:NT3

#A.2 PUD
mydat                <- anno_tot[,match(coloi,colnames(anno_tot))]
ix.distal            <- c(1:nrow(anno_tot))
ix.proximal          <- match(as.character(anno_tot$imIDp),as.character(anno_tot$uniqueID))
psi                  <- mydat
for(i in c(1:ncol(psi))){
  psi[,i]                                                              <- mydat[ix.proximal,i]/(mydat[ix.distal,i]+mydat[ix.proximal,i])
  psi[is.na(psi[,i]),i]<-0
}
PUD                 <- psi
rownames(PUD)       <- anno_tot$uniqueID
colnames(PUD)       <- unlist(lapply(colnames(mydat),function(Z)return(gsub(Z,pattern="\\.raw.corrected",repl=""))))
mPUD                <- t(apply(PUD,1,function(x)return(tapply(x,INDEX=factor(c("NGF.axon","NGF.axon","NT3.axon","NT3.axon")),FUN=mean))))
dPUD                <- mPUD[,1]-mPUD[,2]#NGF:NT3


#A.3 Sum then PUD
mydat                <- anno_tot[,match(coloi,colnames(anno_tot))]
grp                 <- factor(unlist(lapply(FUN=function(Z)return(gsub(Z,pattern="\\.2\\.raw.corrected",repl="")),unlist(lapply(colnames(mydat),function(Z)return(gsub(Z,pattern="\\.1\\.raw.corrected",repl="")))))),levels=c("NGF.axon","NT3.axon")) 
mydat                <- t(apply(mydat,1,function(Z)return(tapply(Z,INDEX=grp,FUN=sum))))
ix.distal            <- c(1:nrow(anno_tot))
ix.proximal          <- match(as.character(anno_tot$imIDp),as.character(anno_tot$uniqueID))
psi                  <- mydat
for(i in c(1:ncol(psi))){
  psi[,i]                                                              <- mydat[ix.proximal,i]/(mydat[ix.distal,i]+mydat[ix.proximal,i])
  psi[is.na(psi[,i]),i]<-0
}

my.proximal <- mydat[ix.proximal,]
my.distal   <- mydat[ix.distal,]
rel.proximal.usage <- cbind(my.proximal[,1]/(my.distal[,1]+my.proximal[,1]),
                            my.proximal[,2]/(my.distal[,2]+my.proximal[,2]))

colnames(rel.proximal.usage) <- c("NGF","NT3")
rownames(rel.proximal.usage)  <- anno_tot$uniqueID
rel.proximal.usage           <- rel.proximal.usage[sela1,]

SUD                 <- psi
rownames(SUD)       <- anno_tot$uniqueID
colnames(SUD)       <- unlist(lapply(colnames(mydat),function(Z)return(gsub(Z,pattern="\\.raw",repl=""))))
dSUD                <- SUD[,1]-SUD[,2]

#B. P-value from Fisher count test
mydat                <- anno_tot[,match(coloi,colnames(anno_tot))]
grp                  <- factor(unlist(lapply(FUN=function(Z)return(gsub(Z,pattern="\\.2\\.raw\\.corrected",repl="")),unlist(lapply(colnames(mydat),function(Z)return(gsub(Z,pattern="\\.1\\.raw\\.corrected",repl="")))))),levels=c("NGF.axon","NT3.axon")) 
mydat                <- t(apply(mydat,1,function(Z)return(tapply(Z,INDEX=grp,FUN=sum))))

test.apa <- function(ix.proximal=ix1[1],ix.distal=ix2[1]){
  return(fisher.test(round(cbind(mydat[ix.proximal,],mydat[ix.distal,])))$p.value)
}
ix1                  <- c(1:nrow(anno_tot))
ix2                  <- match(as.character(anno_tot$imIDp),as.character(anno_tot$uniqueID))
fisherRUD            <- do.call(what=rbind,args=lapply(c(1:nrow(mydat)),function(x)return(test.apa(ix.proximal=ix1[x],ix.distal=ix2[x]))))
padjust.1            <- p.adjust(fisherRUD[sela1],method="fdr")

ix.distal            <- c(1:nrow(anno_tot))
ix.proximal          <- match(as.character(anno_tot$imIDp),as.character(anno_tot$uniqueID))

myE.proximal         <- mydat[ix.proximal,]
colnames(myE.proximal)<-paste(colnames(myE.proximal),".proximal",sep="")
myE.distal           <- mydat[ix.distal,]
colnames(myE.distal)<-paste(colnames(myE.distal),".distal",sep="")
myE                 <- cbind(myE.proximal[,1],myE.distal[,1],myE.proximal[,2],myE.distal[,2])
colnames(myE)       <- c(colnames(myE.proximal)[1],colnames(myE.distal)[1],colnames(myE.proximal)[2],colnames(myE.distal)[2])                      
colnames(mPUD)      <- paste("PUD",colnames(mPUD),sep=".")

myOut               <- data.frame(anno_tot[,c("txID","geneSymbol","imIDp","uniqueID")],dPUD,mPUD,dSUD,SUD,myE)
colnames(myOut)[c(3,4)]<-c("id.proximal","id.distal")


#C. Now identify the differentially used 3' UTR 
id.proxi             <-  anno_tot$imIDp[match(names(dRUD)[sela1],anno_tot$uniqueID)]
seld1                <-  anno_tot$NGF.axon.is.expressed.iso[match(id.proxi,anno_tot$uniqueID)]#Proximal is expressed in NGF
seld2                <-  anno_tot$NGF.axon.is.expressed.iso[match(names(dRUD)[sela1],anno_tot$uniqueID)]#Distal is expressed in NGF
seld3                <-  anno_tot$NT3.axon.is.expressed.iso[match(id.proxi,anno_tot$uniqueID)]#Proximal is expressed in NT3
seld4                <-  anno_tot$NT3.axon.is.expressed.iso[match(names(dRUD)[sela1],anno_tot$uniqueID)]#Distal is expressed in Nt3
sele1                <-  (dPUD>0.15&dSUD>0.15)[sela1]
sele2                <-  (dPUD<(-0.15)&dSUD<(-0.15))[sela1]
self1                <-  dRUD[sela1]<(-1)#NGF-NT3<(-1) i.e. either more of the distal in NGF or more of the proximal in NT3
self2                <-  dRUD[sela1]>(1)#NGF-NT3>1 i.e. either more of the proximal in NGF or more of the distal in NT3
# Significant P-value of difference
selc                 <-  padjust.1<0.01&fisherRUD[sela1]<0.01

mRUD                 <- data.frame(mRUD)
Lim1  <- 0.2
Lim2  <- 0.8

selRUD1             <- list()
IX=1
selRUD1[[IX]]       <- selc&self1&seld2&seld3#498; proximal shift in NT3 compared to NGF so (1) proximal is expressed in NT3 and (2) distal is expressed in NGF
IX=IX+1
selRUD1[[IX]]       <- selc&self2&seld1&seld4#521; proximal shift in NGF compared to NT3 so proximal is expressed in NT3 and distal is expressed in NGF
IX=IX+1
selRUD1[[IX]]      <- selc&self1&seld2&seld3&sele2#310-proximal shift in NT3 so proximal is expressed in axons
IX=IX+1
selRUD1[[IX]]      <- selc&self2&seld1&seld4&sele1#364-proximal shift in NGF so distal expressed in axons

IX=IX+1
selRUD1[[IX]]      <- selc&self1&seld2&seld3&sele2&rel.proximal.usage[,1]<=Lim1
IX=IX+1
selRUD1[[IX]]      <- selc&self2&seld1&seld4&sele1&rel.proximal.usage[,2]>=Lim2


sel.uniqueID.ax  <- lapply(selRUD1,function(Z)return(as.character(anno_tot$uniqueID[sela1])[Z]))
sel.txID.ax      <- lapply(selRUD1,function(Z)return(unique(as.character(anno_tot$txID[sela1])[Z])))
sel.gs.ax        <- lapply(selRUD1,function(Z)return(unique(as.character(anno_tot$geneSymbol[sela1])[Z])))
res.sel.ax       <- lapply(sel.uniqueID.ax,function(Z)return(myOut[match(Z,myOut$id.distal),]))

```